File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
using namespace std;

int main(void) {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n;
  cin >> n;

  while(n--){
    string str;
    cin >> str;

    stack<char> s;
    bool flag = false;
    for(char c:str){
      if(c == '(') s.push(c);
      else{
        if(!s.empty()) s.pop();
        else{
          flag = true;
          break;
        }
      }
    }

    if(!s.empty()) flag = true;

    if(flag) cout << "NO\n";
    else cout << "YES\n";
  }
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        while(n-- > 0){
            String str = br.readLine();

            Stack<Character> s = new Stack<>();
            boolean flag = false;
            for(int i=0; i<str.length(); i++){
                if(str.charAt(i) == '(') s.push(str.charAt(i));
                else {
                    if(!s.empty()) s.pop();
                    else{
                        flag = true;
                        break;
                    }   
                }
            }

            if(!s.empty()) flag = true;

            if(flag) System.out.println("NO");
            else System.out.println("YES");
        }
    }

}
